### DISCUSSION TITLE
Use our applications existing auth with Supabase?

### DISCUSSION THREAD

**Discussion by cory-weiner** [supabase/supabase#21160] [Questions]
I'm looking into supabase at the moment as a solution, primarily for notifications/realtime and presence functionality, but the other capabilities are intruiging.

We are currently using Pusher which allows us to use our existing authentication (basically our internal API signs a token using the Pusher secret key, and that token includes all of the necessary data about the user). Under this pattern, we don't actually create an "users" in Pusher, as they already exist in our system, we just provide the user ID.

I'm wondering if it is possible to accomplish the same thing with Supabase? Looking through the authorization options, it seems Supabase expects an actual user in their database for each user in our system. Is that accurate? Is it not possible to sign a token on our Auth server containing the users credentials, and apply the row level security against the user data in that token?

**Comment by GaryAustin1** [comment]
If you use the Supabase jwt secret (or it works out that you can set it to yours) then you can mint or pass jwt's to Supabase in the Authorization header in createClient or REST calls.  There are a few key claims to use the built in functions, but you can always use auth.jwt() in RLS to get any claim in the jwt.  The use of the user table or gotrue/auth is not required.  To use PostgREST built in roles (authenticated,anon) then you need a roles claim.  But not required as default will be anon if not provided.  Either way your claims can be had with auth.jwt().

https://supabase.com/docs/learn/auth-deep-dive/auth-deep-dive-jwts#jwts-in-supabase
https://supabase.com/docs/guides/realtime/postgres-changes#custom-tokens

**Reply by ahmedivy** [reply]
I am trying to implement that but its not working.

Here I am creating token based on userId from from mongodb:
```ts
export async function createToken() {
  const user = await getUserData();
  const token = await new SignJWT({
    role: "authenticated",
    userid: user.id,
  })
    .setProtectedHeader({ alg: "HS256" })
    .setSubject(user.id)
    .setAudience("authenticated")
    .setIssuedAt()
    .sign(new TextEncoder().encode(process.env.SUPABASE_JWT_SECRET));
  return token;
}
```

Here I created policy on messages table where receiver is a text field that contains userId (from mongodb)
```sql
CREATE POLICY my_policy ON messages
AS PERMISSIVE
FOR SELECT 
TO authenticated
USING ( (auth.jwt() ->> 'userid')::text = receiver );
```

But its not receiving updates in client code.

```ts
"use client";

import { createClient } from "@supabase/supabase-js";
import { useEffect, useState } from "react";

type Messsage = {
  id: number;
  sender: string;
  receiver: string;
  message: string;
  created_at: string;
  read: boolean;
  type: string;
};

export function Chat({ token, chats }: { token: string; chats: Messsage[] }) {
  const [messages, setMessages] = useState<Messsage[]>(chats);

  useEffect(() => {
    const supabase = createClient(
      process.env.NEXT_PUBLIC_SUPABASE_URL!,
      process.env.NEXT_PUBLIC_SUPABASE_ANON_KEY!,
    );

    // Set your custom JWT here
    supabase.realtime.setAuth(token);

    supabase.auth.getSession().then((session) => {
      console.log(session);
    });

    const channel = supabase
      .channel("db-changes")
      .on(
        "postgres_changes",
        {
          event: "*",
          schema: "public",
          table: "messages",
        },
        (payload: any) => {
          console.log("Change received!", payload);
          if (payload.eventType === "INSERT") {
            setMessages((prev) => [...prev, payload.new]);
          }
        },
      )
      .subscribe();

    supabase
      .from("messages")
      .select("*")
      .then(({ data, error }) => {
        console.log(data, error);
      });

    return () => {
      supabase.removeChannel(channel);
    };
  }, [token]);

  return (
    <main>
      <ul>
        {messages.map((message) => (
          <li key={message.id}>
            <p>{message.message}</p>
            <p>{message.sender}</p>
            <p>{message.created_at}</p>
            <p>---------------------------------</p>
          </li>
        ))}
      </ul>
    </main>
  );
}
```

**Reply by GaryAustin1** [reply]
You must be using old supabase-js code.  There is no setAuth() command anymore.  You set the Authorization header in the createClient call with the header option.  Note there will be no session so you can use any session calls with your own jwt.
https://supabase.com/docs/reference/javascript/initializing?example=with-additional-parameters

**Reply by ahmedivy** [reply]
Thanks @GaryAustin1,

Its still not sending payload when I create a new record from supabase dashboard, but if I explicitly fetch records with same client using `supabase.from("messages").select("*");` it works, I dont know why its not sending payload onChange. Realtime is enabled for this table.

This is my code now
```ts
"use client";

import { createClient } from "@supabase/supabase-js";
import { useEffect, useState } from "react";

type Messsage = {
  id: number;
  sender: string;
  receiver: string;
  message: string;
  created_at: string;
  read: boolean;
  type: string;
};

export function Chat({ token, chats }: { token: string; chats: Messsage[] }) {
  const [messages, setMessages] = useState<Messsage[]>(chats);

  useEffect(() => {
    const supabase = createClient(
      process.env.NEXT_PUBLIC_SUPABASE_URL!,
      process.env.NEXT_PUBLIC_SUPABASE_ANON_KEY!,
      {
        global: {
          headers: {
            Authorization: `Bearer ${token}`,
          },
        },
      },
    );

    const channel = supabase
      .channel("db-changes")
      .on(
        "postgres_changes",
        {
          event: "*",
          schema: "public",
          table: "messages",
        },
        (payload: any) => {
          // this function not called when i create new record from supabase dashboard (satisfying rls policy)
          console.log("Change received!", payload);
          if (payload.eventType === "INSERT") {
            setMessages((prev) => [...prev, payload.new]);
          }
        },
      )
      .subscribe();

    supabase
      .from("messages")
      .select("*")
      .then(({ data }) => {
        console.log(data); // here i receive the data
        setMessages(data as Messsage[]);
      });

    return () => {
      supabase.removeChannel(channel);
    };
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  return (
    <main>
      <ul>
        {messages.map((message) => (
          <li key={message.id}>
            <p>{message.message}</p>
            <p>{message.sender}</p>
            <p>{message.created_at}</p>
            <p>---------------------------------</p>
          </li>
        ))}
      </ul>
    </main>
  );
}
```

**Reply by GaryAustin1** [reply]
OK I missed the realtime.setAuth().  Sorry for the confusion.  There use to be a plain setAuth() that was used to set the session to a JWT for all of the supabase client.  You still do need the realtime.setAuth().  Follow the docs for that, I've not personally done it.

**Reply by ahmedivy** [reply]
In [docs](https://supabase.com/docs/guides/realtime/postgres-changes#custom-tokens), they said to only call `supabase.realtime.setAuth()`. That's what I was trying first. Now I've tried both methods (by setting headers in the options field and calling this function), but it's still not working.

If I just fetch the data from same client it works, so I think jwt is correct.

**Reply by GaryAustin1** [reply]
The realtime.setAuth() only applies to realtime. So yes you need both.  
You can add .subscribe((status)=>console.log(status)) to see if there are any errors, or check the realtime logs.   You can also use the reatime tab on the dashboard to see if the messages are being sent.  Or this: https://realtime.supabase.co/inspector/new .

Also just linking another similar thing from awhile ago https://github.com/supabase/supabase/issues/8743
